zerver/openapi/curl_examples.py

```diff
@@ -0,0 +1,73 @@
+from typing import Dict, Any, Optional, Callable, Set, List
+
+from functools import wraps
+
+from zerver.lib.actions import check_send_stream_message, bulk_add_subscriptions
+from zerver.models import get_realm, get_user, Client, Realm, get_stream
+
+TEST_FUNCTIONS = dict()  # type: Dict[str, Callable[..., None]]
+REGISTERED_TEST_FUNCTIONS = set()  # type: Set[str]
+CALLED_TEST_FUNCTIONS = set()  # type: Set[str]
+
+def openapi_test_function(endpoint: str) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+    """This decorator is used to register an openapi test function with
+    its endpoint. Example usage:
+
+    @openapi_test_function("/messages/render:post")
+    def ...
+    """
+    def wrapper(test_func: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(test_func)
+        def _record_calls_wrapper(*args: Any, **kwargs: Any) -> Any:
+            CALLED_TEST_FUNCTIONS.add(test_func.__name__)
+            return test_func(*args, **kwargs)
+
+        REGISTERED_TEST_FUNCTIONS.add(test_func.__name__)
+        TEST_FUNCTIONS[endpoint] = _record_calls_wrapper
+
+        return _record_calls_wrapper
+    return wrapper
+
+def send_stream_message(sender_email: str, stream_name: str, content: str="test content",
+                        topic_name: str="test", sender_realm: str="zulip",
+                        recipient_realm: Optional[Realm]=None) -> int:
+    sender = get_user(sender_email, get_realm(sender_realm))
+
+    (sending_client, _) = Client.objects.get_or_create(name="test suite")
+
+    return check_send_stream_message(
+        sender=sender,
+        client=sending_client,
+        stream_name=stream_name,
+        topic=topic_name,
+        body=content,
+        realm=recipient_realm,
+    )
+
+@openapi_test_function("/messages/{message_id}:get")
+@openapi_test_function("/messages/{message_id}/history:get")
+def iago_message_id() -> Dict[str, int]:
+    return {
+        "message_id": send_stream_message("iago@zulip.com", "Denmark")
+    }
```

Comment:
This might be better done with `@openapi_test_function` taking *args arguments and just listing the endpoints there?  Otherwise we'll potentially end up with a pretty deep nesting here.